neo/Consensus/RecoveryMessage.cs

```diff
@@ -0,0 +1,126 @@
+﻿using System;
+using System.IO;
+using System.Linq;
+using Neo.IO;
+using Neo.Network.P2P.Payloads;
+
+namespace Neo.Consensus
+{
+    internal class RecoveryMessage : ConsensusMessage
```

Comment:
@erikzhang, maybe #426 is something for us to reconsider. This payload is quite complex compared to the solution we used there.

Furthermore, the design used there can proportionate us more field for field improvements on P2P messages communication and optimization in the future, for a analyzing a trade-off between payload complexity and communication efficiency.

Reply:
There is one major advantage I see of this code over the other code. If the code changes view to a higher view number but on that view a `PrepareRequest` hasn't even yet been sent; this code will still regenerate a node that restarts, so that it is brought back to the current view number. I don't believe #426 handles that. One of the reasons for the additional complexity in this payload is to handle that case and cause the payload not to inclkude the unnecessary fields when a `PrepareRequest` hasn't been sent yet.

Reply:
We use the message itself to force the view.

Reply:
@vncoelho Yeah, but it won't even send the message in #426 if the request wasn't sent yet in the view. This code will still send the RecoveryMessage even if no request was sent yet in the view.

Reply:
It will send the Recovery as soon as any ChangeView payload is sent.

Reply:
That is true, Jeff.
Are there another cases covered here? We already merged the hardware recover that Erik implemented with the Log.

Reply:
This covers recovering the view changes, the prepare request messages, and the prepare response messages; everything required for a node to jump immediately back to the current state of other consensus nodes.

Reply:
Now it also is also smarter in recovery and if enough preparations are present in the recovery message, it doesn't need to regenerate the view messages.

Reply:
So while this payload seems complex, the complexity is necessary in order to be able to regenerate all 3 types of messages (`ChangeView`, `PrepareRequest`, and `PrepareResponse`). The class is smart so that it doesn't include information not needed in order to have a smaller payload when serialized based on the situation of what is needed in the recovery message. The complexity is warranted for the benefits gained in ability to recover better.

Is there any reason we cannot resolve this conversation now?

Reply:
The requirements are not yet accomplished.
When those cases are covered we will double check here to see if all data from the payload are used to cover fault cases, as well as if something else can be simplicity as in the previous PR.

Reply:
After talking offline, we discussed one other fault case that I think you are referring to. I will adjust so the primary can recover into a view with commit sent and re-accept his previously sent `PrepareRequest` along with the change I mention in the discussion below that allows recovery to accept from a lower view number if there are enough prepare requests signed in the payload for him to immediately commit in the lower view.

Reply:
I think your last conversation now covers the the only remaining thing on this conversation, so this one could be closed.